File: examples/tracking/multi-camera-people-detection-and-tracking-insight/cpp/utils/pipeline.cpp
#include "examples/tracking/multi-camera-people-detection-and-tracking-insight/cpp/utils/pipeline_api.cpp"

#include "nodes/groups/VideoSender.h"

#include <opencv2/videoio.hpp>

#include <cmath>
#include <memory>
#include <optional>
#include <stdexcept>
#include <vector>

namespace multi_camera_people_tracking {
namespace {

constexpr double kYoloV8DefaultDetectionThreshold = 0.6;
constexpr double kYoloV8DefaultNmsIouThreshold = 0.5;
constexpr int kYoloV8DefaultTopK = 24;

} // namespace

simaai::neat::Sample run_sample_input_once(simaai::neat::Run& run,
                                           const simaai::neat::Sample& input, int timeout_ms) {
  auto outputs = run.run(simaai::neat::SampleList{input}, timeout_ms);
  if (outputs.empty()) {
    throw std::runtime_error("detector run produced no samples");
  }
  return std::move(outputs.front());
}

int insight_video_port_for_stream(int port_base, int stream_index) {
  return port_base + stream_index;
}

int insight_metadata_port_for_stream(int port_base, int stream_index) {
  return port_base + stream_index;
}

int effective_writer_fps(const AppConfig& cfg, const RtspProbe& probe) {
  return cfg.fps > 0 ? cfg.fps : std::max(probe.fps, 10);
}

RtspProbe probe_rtsp(const std::string& url) {
  cv::VideoCapture capture(url);
  if (!capture.isOpened()) {
    throw std::runtime_error("failed to open RTSP source for probing: " + url);
  }

  const int width = static_cast<int>(capture.get(cv::CAP_PROP_FRAME_WIDTH));
  const int height = static_cast<int>(capture.get(cv::CAP_PROP_FRAME_HEIGHT));
  const int fps = static_cast<int>(std::lround(capture.get(cv::CAP_PROP_FPS)));
  capture.release();

  if (width <= 0 || height <= 0) {
    throw std::runtime_error("failed to probe RTSP frame size: " + url);
  }
  return RtspProbe{width, height, std::max(0, fps)};
}

SessionRun build_source_run(const AppConfig& cfg, const std::string& url, const RtspProbe& probe) {
  simaai::neat::nodes::groups::RtspDecodedInputOptions options;
  options.url = url;
  options.latency_ms = cfg.latency_ms;
  options.tcp = cfg.tcp;
  options.payload_type = 96;
  options.insert_queue = true;
  options.out_format = "RGB";
  options.decoder_raw_output = false;
  options.auto_caps_from_stream = true;
  options.use_videoconvert = false;
  options.use_videoscale = true;
  options.fallback_h264_width = probe.width;
  options.fallback_h264_height = probe.height;
  if (probe.fps > 0) {
    options.fallback_h264_fps = probe.fps;
  }
  options.output_caps.enable = true;
  options.output_caps.format = "RGB";
  options.output_caps.width = probe.width;
  options.output_caps.height = probe.height;
  if (probe.fps > 0) {
    options.output_caps.fps = probe.fps;
  }
  options.output_caps.memory = simaai::neat::CapsMemory::SystemMemory;

  SessionRun runtime;
  runtime.session.add(simaai::neat::nodes::groups::RtspDecodedInput(options));
  runtime.session.add(simaai::neat::nodes::Output(simaai::neat::OutputOptions::EveryFrame(1)));

  simaai::neat::RunOptions run_options;
  run_options.queue_depth = 4;
  run_options.overflow_policy = simaai::neat::OverflowPolicy::KeepLatest;
  run_options.output_memory = simaai::neat::OutputMemory::Owned;
  runtime.run = runtime.session.build(run_options);
  return runtime;
}

SessionRun build_detection_run(const AppConfig& cfg, const RtspProbe& probe) {
  SessionRun runtime;

  simaai::neat::Model::Options model_options;
  model_options.preprocess.kind = simaai::neat::InputKind::Image;
  model_options.preprocess.color_convert.input_format = simaai::neat::PreprocessColorFormat::RGB;
  model_options.preprocess.input_max_width = probe.width;
  model_options.preprocess.input_max_height = probe.height;
  model_options.preprocess.input_max_depth = 3;
  model_options.decode_type = simaai::neat::BoxDecodeType::YoloV8;
  model_options.score_threshold =
      cfg.detection_threshold.value_or(kYoloV8DefaultDetectionThreshold);
  model_options.nms_iou_threshold = cfg.nms_iou_threshold.value_or(kYoloV8DefaultNmsIouThreshold);
  model_options.top_k = cfg.top_k.value_or(kYoloV8DefaultTopK);
  model_options.boxdecode_original_width = probe.width;
  model_options.boxdecode_original_height = probe.height;
  runtime.model = std::make_shared<simaai::neat::Model>(cfg.model, model_options);

  auto input_options = runtime.model->input_appsrc_options(false);
  input_options.media_type = "video/x-raw";
  input_options.format = "RGB";
  input_options.width = probe.width;
  input_options.height = probe.height;
  input_options.depth = 3;
  runtime.session.add(simaai::neat::nodes::Input(input_options));
  runtime.session.add(simaai::neat::nodes::groups::Preprocess(*runtime.model));
  runtime.session.add(simaai::neat::nodes::groups::Infer(*runtime.model));
  runtime.session.add(simaai::neat::nodes::SimaBoxDecode(
      *runtime.model, simaai::neat::BoxDecodeType::YoloV8,
      cfg.detection_threshold.value_or(kYoloV8DefaultDetectionThreshold),
      cfg.nms_iou_threshold.value_or(kYoloV8DefaultNmsIouThreshold),
      cfg.top_k.value_or(kYoloV8DefaultTopK), "", std::nullopt, std::nullopt, probe.width,
      probe.height));
  runtime.session.add(simaai::neat::nodes::Output());

  cv::Mat seed = cv::Mat::zeros(probe.height, probe.width, CV_8UC3);
  simaai::neat::RunOptions run_options;
  run_options.preset = simaai::neat::RunPreset::Realtime;
  run_options.queue_depth = 1;
  run_options.overflow_policy = simaai::neat::OverflowPolicy::KeepLatest;
  run_options.output_memory = simaai::neat::OutputMemory::Owned;
  runtime.run =
      runtime.session.build(std::vector<cv::Mat>{seed}, simaai::neat::RunMode::Async, run_options);
  return runtime;
}

SessionRun build_insight_video_run(const AppConfig& cfg, const RtspProbe& probe, int stream_index) {
  simaai::neat::InputOptions input_options;
  input_options.media_type = "video/x-raw";
  input_options.format = "RGB";
  input_options.use_simaai_pool = false;
  input_options.max_width = probe.width;
  input_options.max_height = probe.height;
  input_options.max_depth = 3;

  SessionRun runtime;
  runtime.session.add(simaai::neat::nodes::Input(input_options));
  auto video_options = simaai::neat::nodes::groups::VideoSenderOptions::H264RtpUdpFromRaw(
      probe.width, probe.height, effective_writer_fps(cfg, probe));
  video_options.host = cfg.insight_host;
  video_options.channel = stream_index;
  video_options.video_port_base = cfg.insight_video_port_base;
  video_options.sync = false;
  video_options.async = false;
  video_options.rtp.payload_type = 96;
  video_options.rtp.config_interval = 1;
  video_options.encoder.bitrate_kbps = cfg.bitrate_kbps;
  video_options.encoder.profile = "baseline";
  video_options.encoder.level = "4.1";
  runtime.session.add(simaai::neat::nodes::groups::VideoSender(video_options));

  cv::Mat seed = cv::Mat::zeros(probe.height, probe.width, CV_8UC3);
  simaai::neat::RunOptions run_options;
  run_options.queue_depth = 2;
  run_options.overflow_policy = simaai::neat::OverflowPolicy::KeepLatest;
  runtime.run =
      runtime.session.build(std::vector<cv::Mat>{seed}, simaai::neat::RunMode::Async, run_options);
  return runtime;
}

simaai::neat::MetadataSender build_insight_metadata_output(const AppConfig& cfg, int stream_index) {
  simaai::neat::MetadataSenderOptions options;
  options.host = cfg.insight_host;
  options.channel = stream_index;
  options.metadata_port_base = cfg.insight_metadata_port_base;
  return simaai::neat::MetadataSender(options);
}

} // namespace multi_camera_people_tracking

File: examples/tracking/multi-camera-people-detection-and-tracking-insight/python/utils/pipeline.py
"""Pipeline builders and runtime helpers for the Python multi-camera example."""

from __future__ import annotations

from dataclasses import dataclass
import glob
import sys
from typing import Any

from .config import AppConfig


@dataclass(frozen=True)
class RtspProbe:
    width: int
    height: int
    fps: int


@dataclass(frozen=True)
class RuntimeModules:
    cv2: Any
    np: Any
    pyneat: Any


@dataclass(frozen=True)
class QuantTessCpuPreproc:
    width: int
    height: int
    aspect_ratio: bool
    padding_type: str


_RUNTIME_MODULES: RuntimeModules | None = None

_YOLOV8_BOXDECODE_DEFAULTS = {
    "detection_threshold": 0.6,
    "nms_iou_threshold": 0.5,
    "topk": 24,
}

_SOURCE_STARTUP_PULL_TIMEOUT_MS = 50000
_SOURCE_PULL_TIMEOUT_MS = 10000
_SOURCE_STARTUP_STAGGER_S = 0.5


def insight_video_port_for_stream(port_base: int, stream_index: int) -> int:
    return int(port_base) + int(stream_index)


def insight_metadata_port_for_stream(port_base: int, stream_index: int) -> int:
    return int(port_base) + int(stream_index)


def effective_writer_fps(cfg: AppConfig, probe: RtspProbe) -> int:
    return cfg.fps if cfg.fps > 0 else max(probe.fps, 10)


def load_runtime_modules() -> RuntimeModules:
    global _RUNTIME_MODULES
    if _RUNTIME_MODULES is not None:
        return _RUNTIME_MODULES

    for path in glob.glob("/usr/lib/python3*/dist-packages"):
        if path not in sys.path:
            sys.path.insert(0, path)

    try:
        import cv2
        import numpy as np
        import pyneat
    except ModuleNotFoundError as exc:
        raise RuntimeError(
            "runtime dependencies are missing; install numpy, OpenCV, and pyneat "
            "or run inside the NEAT Python environment"
        ) from exc

    _RUNTIME_MODULES = RuntimeModules(cv2=cv2, np=np, pyneat=pyneat)
    return _RUNTIME_MODULES


def probe_rtsp(url: str) -> RtspProbe:
    runtime = load_runtime_modules()
    cap = runtime.cv2.VideoCapture(url)
    if not cap.isOpened():
        raise RuntimeError(f"failed to open RTSP source for probing: {url}")
    width = int(cap.get(runtime.cv2.CAP_PROP_FRAME_WIDTH) or 0)
    height = int(cap.get(runtime.cv2.CAP_PROP_FRAME_HEIGHT) or 0)
    fps = int(round(cap.get(runtime.cv2.CAP_PROP_FPS) or 0))
    cap.release()
    if width <= 0 or height <= 0:
        raise RuntimeError(f"failed to probe RTSP frame size: {url}")
    return RtspProbe(width=width, height=height, fps=max(0, fps))


def load_detector_model(runtime: RuntimeModules, cfg: AppConfig):
    # Load the model pack in tensor mode because CPU preproc produces the
    # FP32 image tensor consumed by QuantTess in the explicit detector graph.
    pyneat = runtime.pyneat
    opt = pyneat.ModelOptions()
    opt.media_type = "application/vnd.simaai.tensor"
    opt.format = ""
    return pyneat.Model(cfg.model, opt)


def read_preproc_contract(runtime: RuntimeModules, model: Any) -> QuantTessCpuPreproc:
    # Reuse the packaged preproc geometry even though preprocessing happens on CPU.
    pyneat = runtime.pyneat
    pre = pyneat.PreprocOptions(model)
    cfg_json = dict(getattr(pre, "config_json", None) or {})
    width = int(cfg_json.get("output_width") or cfg_json.get("input_width") or 640)
    height = int(cfg_json.get("output_height") or cfg_json.get("input_height") or 640)
    return QuantTessCpuPreproc(
        width=width,
        height=height,
        aspect_ratio=bool(cfg_json.get("aspect_ratio", False)),
        padding_type=str(cfg_json.get("padding_type", "CENTER")).upper(),
    )


def build_source_run(runtime: RuntimeModules, cfg: AppConfig, url: str, probe: RtspProbe):
    pyneat = runtime.pyneat
    ro = pyneat.RtspDecodedInputOptions()
    ro.url = url
    ro.latency_ms = cfg.latency_ms
    ro.tcp = cfg.tcp
    ro.payload_type = 96
    ro.insert_queue = True
    ro.auto_caps_from_stream = True
    ro.sima_allocator_type = 2
    ro.decoder_raw_output = False
    ro.use_videoconvert = False
    ro.use_videoscale = True
    ro.fallback_h264_width = probe.width
    ro.fallback_h264_height = probe.height
    if probe.fps > 0:
        ro.fallback_h264_fps = probe.fps
    ro.output_caps.enable = True
    ro.output_caps.width = probe.width
    ro.output_caps.height = probe.height
    if probe.fps > 0:
        ro.output_caps.fps = probe.fps
    ro.output_caps.memory = pyneat.CapsMemory.SystemMemory

    session = pyneat.Session()
    session.add(pyneat.groups.rtsp_decoded_input(ro))
    session.add(pyneat.nodes.output(pyneat.OutputOptions.every_frame(1)))
    run_opt = pyneat.RunOptions()
    run_opt.queue_depth = 4
    run_opt.overflow_policy = pyneat.OverflowPolicy.KeepLatest
    run_opt.output_memory = pyneat.OutputMemory.Owned
    run = session.build(run_opt)
    return session, run


def build_detection_run(
    runtime: RuntimeModules,
    cfg: AppConfig,
    probe: RtspProbe,
):
    pyneat = runtime.pyneat
    np = runtime.np

    model_opt = pyneat.ModelOptions()
    model_opt.preprocess.kind = pyneat.InputKind.Image
    model_opt.preprocess.color_convert.input_format = pyneat.PreprocessColorFormat.RGB
    model_opt.preprocess.input_max_width = probe.width
    model_opt.preprocess.input_max_height = probe.height
    model_opt.preprocess.input_max_depth = 3
    model_opt.decode_type = pyneat.BoxDecodeType.YoloV8
    model_opt.score_threshold = (
        cfg.detection_threshold
        if cfg.detection_threshold is not None
        else _YOLOV8_BOXDECODE_DEFAULTS["detection_threshold"]
    )
    model_opt.nms_iou_threshold = (
        cfg.nms_iou_threshold
        if cfg.nms_iou_threshold is not None
        else _YOLOV8_BOXDECODE_DEFAULTS["nms_iou_threshold"]
    )
    model_opt.top_k = (
        cfg.top_k if cfg.top_k is not None else _YOLOV8_BOXDECODE_DEFAULTS["topk"]
    )
    model_opt.boxdecode_original_width = probe.width
    model_opt.boxdecode_original_height = probe.height
    model = pyneat.Model(cfg.model, model_opt)

    input_opt = model.input_appsrc_options(False)
    input_opt.media_type = "video/x-raw"
    input_opt.format = "RGB"
    input_opt.width = probe.width
    input_opt.height = probe.height
    input_opt.depth = 3
    for attr, value in (
        ("max_width", probe.width),
        ("max_height", probe.height),
        ("max_depth", 3),
    ):
        if hasattr(input_opt, attr):
            setattr(input_opt, attr, value)

    session = pyneat.Session()
    session.add(pyneat.nodes.input(input_opt))
    session.add(model.preprocess())
    session.add(pyneat.groups.mla(model))
    session.add(
        pyneat.nodes.sima_box_decode(
            model,
            decode_type=pyneat.BoxDecodeType.YoloV8,
            original_width=probe.width,
            original_height=probe.height,
            detection_threshold=(
                cfg.detection_threshold
                if cfg.detection_threshold is not None
                else _YOLOV8_BOXDECODE_DEFAULTS["detection_threshold"]
            ),
            nms_iou_threshold=(
                cfg.nms_iou_threshold
                if cfg.nms_iou_threshold is not None
                else _YOLOV8_BOXDECODE_DEFAULTS["nms_iou_threshold"]
            ),
            top_k=cfg.top_k if cfg.top_k is not None else _YOLOV8_BOXDECODE_DEFAULTS["topk"],
        )
    )
    session.add(pyneat.nodes.output())

    seed = pyneat.Tensor.from_numpy(
        np.zeros((probe.height, probe.width, 3), dtype=np.uint8),
        copy=True,
        image_format=pyneat.PixelFormat.RGB,
        memory=pyneat.TensorMemory.EV74,
    )
    run_opt = pyneat.RunOptions()
    run_opt.preset = pyneat.RunPreset.Realtime
    run_opt.queue_depth = 1
    run_opt.overflow_policy = pyneat.OverflowPolicy.KeepLatest
    run_opt.output_memory = pyneat.OutputMemory.Owned
    run = session.build(seed, pyneat.RunMode.Async, run_opt)
    return session, run


def build_insight_video_run(
    runtime: RuntimeModules,
    cfg: AppConfig,
    probe: RtspProbe,
    stream_index: int,
):
    pyneat = runtime.pyneat
    np = runtime.np

    input_opt = pyneat.InputOptions()
    input_opt.media_type = "video/x-raw"
    input_opt.format = "RGB"
    input_opt.use_simaai_pool = False
    input_opt.max_width = probe.width
    input_opt.max_height = probe.height
    input_opt.max_depth = 3

    session = pyneat.Session()
    session.add(pyneat.nodes.input(input_opt))
    sender_opt = pyneat.VideoSenderOptions.h264_rtp_udp_from_raw(
        probe.width,
        probe.height,
        effective_writer_fps(cfg, probe),
    )
    sender_opt.host = cfg.insight_host
    sender_opt.channel = stream_index
    sender_opt.video_port_base = cfg.insight_video_port_base
    sender_opt.sync = False
    sender_opt.async_ = False
    sender_opt.rtp.payload_type = 96
    sender_opt.rtp.config_interval = 1
    sender_opt.encoder.bitrate_kbps = cfg.bitrate_kbps
    sender_opt.encoder.profile = "baseline"
    sender_opt.encoder.level = "4.1"
    session.add(pyneat.groups.video_sender(sender_opt))

    seed = pyneat.Tensor.from_numpy(
        np.zeros((probe.height, probe.width, 3), dtype=np.uint8),
        copy=True,
        image_format=pyneat.PixelFormat.RGB,
        memory=pyneat.TensorMemory.EV74,
    )
    run_opt = pyneat.RunOptions()
    run_opt.queue_depth = 2
    run_opt.overflow_policy = pyneat.OverflowPolicy.KeepLatest
    run = session.build(seed, pyneat.RunMode.Async, run_opt)
    return session, run


def build_insight_metadata_output(
    runtime: RuntimeModules,
    cfg: AppConfig,
    stream_index: int,
):
    pyneat = runtime.pyneat
    options = pyneat.MetadataSenderOptions()
    options.host = cfg.insight_host
    options.channel = stream_index
    options.metadata_port_base = cfg.insight_metadata_port_base
    return pyneat.MetadataSender(options)
